### PR TITLE
Detect avx512 in test infra to cover more platforms

### PR DESCRIPTION
**Pull request by am11**
https://github.com/dotnet/runtime/issues/105472#issuecomment-2266377391

### CODE REVIEW

**Comment by am11** [comment]
@Thefrank this should fix the issue but I haven't tested it yet. Could you please rerun your pipeline with smoke tests?

File: src/tests/nativeaot/SmokeTests/HardwareIntrinsics/X64Avx512.csproj [hunk 1]
```diff
@@ -18,11 +18,9 @@
   <PropertyGroup>
     <CLRTestBashPreCommands><![CDATA[
 $(CLRTestBashPreCommands)
-    if [[ "$OSTYPE" == "linux"* ]]; then
-        if ! grep -q '^flags.*avx512' /proc/cpuinfo 2>/dev/null; then
-          echo No support for AVX512, test not applicable.
-          exit 0
-        fi
+    if ! cc -dM -E -march=native - < /dev/null | grep -q __AVX512; then
```
Comment (am11) [inline src/tests/nativeaot/SmokeTests/HardwareIntrinsics/X64Avx512.csproj:21]:
```suggestion
    if ! cc -dM -E -march=native - < /dev/null 2>/dev/null | grep -q __AVX512; then
```

File: src/tests/nativeaot/SmokeTests/HardwareIntrinsics/X64Avx512.csproj [hunk 2]
```diff
@@ -18,11 +18,9 @@
   <PropertyGroup>
     <CLRTestBashPreCommands><![CDATA[
 $(CLRTestBashPreCommands)
-    if [[ "$OSTYPE" == "linux"* ]]; then
-        if ! grep -q '^flags.*avx512' /proc/cpuinfo 2>/dev/null; then
```
Comment (jkotas) [inline src/tests/nativeaot/SmokeTests/HardwareIntrinsics/X64Avx512.csproj:22]: Similar `grep` commands are used in number of other .csproj files. Do all of them need the same fix?

**Comment by jkotas** [comment]
Is the problem that the avx512 string is in `/proc/cpuinfo` on FreeBSD even when the machine does not support avx512?

**Comment by am11** [comment]
> Is the problem that the avx512 string is in /proc/cpuinfo on FreeBSD even when the machine does not support avx512?

It doesn't exist on FreeBSD and illumos etc. text-based procfs is linux specific.

**Review by MichalStrehovsky — APPROVED** [review]
Nice, thanks!

**Comment by Thefrank** [comment]
On the FreeBSD end:
```
Time Elapsed 00:00:29.85
Test run finished.
Parsing test results from (/root/runtime/artifacts/log)
Analyzing /root/runtime/artifacts/log/TestRun.xml
Time [secs] | Total | Passed | Failed | Skipped | Assembly Execution Summary
============================================================================
      0.362 |     1 |      1 |      0 |       0 | nativeaot.CustomMain.XUnitWrapper.dll
      0.085 |     1 |      1 |      0 |       0 | nativeaot.CustomMainWithStubExe.XUnitWrapper.dll
      0.074 |     1 |      1 |      0 |       0 | nativeaot.GenerateUnmanagedEntryPoints.XUnitWrapper.dll
     25.219 |    20 |     20 |      0 |       0 | nativeaot.SmokeTests.XUnitWrapper.dll
      0.039 |     1 |      1 |      0 |       0 | nativeaot.StartupHook.XUnitWrapper.dll
----------------------------------------------------------------------------
     25.779 |    24 |     24 |      0 |       0 | (total)


Log files at: /root/runtime/artifacts/log
```

**Pull request merged**